input/tar2/tar_input.go

```diff
@@ -0,0 +1,163 @@
+package tar2
+
+import (
+	"archive/tar"
+	"bytes"
+	"crypto/sha512"
+	"encoding/base64"
+	"hash"
+	"io"
+	"os"
+	"path"
+
+	"github.com/spacemonkeygo/errors"
+	"github.com/spacemonkeygo/errors/try"
+	"polydawn.net/repeatr/def"
+	"polydawn.net/repeatr/input"
+	"polydawn.net/repeatr/lib/fs"
+	"polydawn.net/repeatr/lib/fshash"
+	"polydawn.net/repeatr/lib/treewalk"
+)
+
+const Type = "tar"
+
+var _ input.Input = &Input{} // interface assertion
+
+type Input struct {
+	spec          def.Input
+	hasherFactory func() hash.Hash
+}
+
+func New(spec def.Input) input.Input {
+	if spec.Type != Type {
+		panic(errors.ProgrammerError.New("This input implementation supports definitions of type %q, not %q", Type, spec.Type))
+	}
+	_, err := os.Stat(spec.URI)
+	if os.IsNotExist(err) {
+		panic(input.DataSourceUnavailableError.New("Input URI %q must be a tar file", spec.URI))
+	}
+	return &Input{
+		spec:          spec,
+		hasherFactory: sha512.New384,
+	}
+}
+
+func (i Input) Apply(destinationRoot string) <-chan error {
+	done := make(chan error)
+	go func() {
+		defer close(done)
+		try.Do(func() {
+			// do make a dir for untaring into.
+			// tars may specify permission and time bits for their top dir, but if not, we'll start with sane defaults.
+			err := os.MkdirAll(destinationRoot, 0755)
+			if err != nil {
+				panic(input.TargetFilesystemUnavailableIOError(err))
+			}
+			// daemon uid and gid are fine for now (they're always 0);
+			// forcing a constant time happens along with all the other dirs, since
+			// the bucket normalizes to include a root attribute set if the tar doesn't have one.
+
+			// open the tar file; preparing decompression as necessary
+			file, err := os.OpenFile(i.spec.URI, os.O_RDONLY, 0755)
+			if err != nil {
+				panic(input.DataSourceUnavailableIOError(err))
+			}
+			defer file.Close()
+			reader, err := Decompress(file)
+			if err != nil {
+				panic(input.DataSourceUnavailableError.New("could not start decompressing: %s", err))
+			}
+			tarReader := tar.NewReader(reader)
+
+			// unroll the tar, copying and accumulating data for integrity check
+			bucket := &fshash.MemoryBucket{}
+			walk(tarReader, destinationRoot, bucket, i.hasherFactory)
+
+			// hash whole tree
+			actualTreeHash, _ := fshash.Hash(bucket, i.hasherFactory)
+
+			// verify total integrity
+			expectedTreeHash, err := base64.URLEncoding.DecodeString(i.spec.Hash)
```

Comment:
Cool link. 

I actually think you're combining two problems into one here: I expect to run into same-hash-type comparison challenges before I expect to manage a full-blown multi-hash cross-compatibility system. Both are inevitable; I assert the later is far larger & further down the road.

Compatibility across hash algorithms would require stuff like telling an output that yeah, actually we need an MD5 for ~reasons~, and stuff like that. Contrast with what I'm looking for, which is merely "have a clearly defined way to represent hashes outside repeatr".

Which is a long-winded way of saying: I'm fine with treating repeatr hashes as opaque utf8 strings for now, and tackling hash representation later & before tackling cross-hash stuff. Multihash looks like a good starting point...